### PR TITLE
feat: Add MERGE DDL

### LINKED ISSUE #428: Add support for MERGE DDL
**Is your feature request related to a problem? Please describe.**
Yes, programmatically generating MERGE statements.

 **Describe the solution you'd like**
Adding support `MERGE` DDL operation in BigQuery.

https://cloud.google.com/bigquery/docs/reference/standard-sql/dml-syntax#merge_statement

 **Describe alternatives you've considered**
Directly writing SQL code

 **Additional context**
None

### LINKED ISSUE #15800: Add support for MERGE DDL
**Is your feature request related to a problem? Please describe.**
Yes, programmatically generating MERGE statements.

 **Describe the solution you'd like**
Adding support `MERGE` DDL operation in BigQuery.

https://cloud.google.com/bigquery/docs/reference/standard-sql/dml-syntax#merge_statement

 **Describe alternatives you've considered**
Directly writing SQL code

 **Additional context**
None

### PR DESCRIPTION
**Pull request by munro**
## help wanted!! ⛑

Throwing over some code I'm currently using to generate `MERGE` statements.

Overall I really like that API, and it feels really slick when I've been using it.

One thing that is missing is `ThenUpdate` / `ThenInsert` doesn't check if the key actually exists on the `into` table—which may be desirable in the instance the developer didn't describe the table in SQLAlchemy (I often use it like that).  But it would be nice to check if the column does exist if the schema is defined.

Fixes googleapis/google-cloud-python#15800 🦕🙌🆒

### CODE REVIEW

**Comment by tswast** [comment]
@munro Thanks so much for the contribution. I hope to review soon. Unfortunately, I cannot merge this without the account associated with your commit (I see a kw.com email) signing the Google CLA at https://cla.developers.google.com/clas

**Review by tswast — CHANGES_REQUESTED** [review]
Very cool! A few initial comments, though the CLA issue will also need to be resolved.

**Comment by pykenny** [comment]
Any updates to this PR?

Also, merge statement in BigQuery allows more complicated case branching like:
```sql
MERGE INTO target AS t
USING source AS s
ON t.id == s.id
-- Special cases
WHEN MATCHED AND s.expire_at <= CURRENT_DATETIME() THEN DELETE
WHEN NOT MATCHED BY SOURCE AND t.expire_at <= CURRENT_DATETIME() THEN DELETE
WHEN NOT MATCHED AND (s.flag_a == 2) THEN INSERT ...
-- Defaults
WHEN MATCHED THEN UPDATE ...
WHEN NOT MATCHED BY SOURCE THEN UPDATE SET ...
WHEN NOT MATCHED THEN INSERT ...
-- Unreachable cases
WHEN MATCHED THEN UPDATE ...
WHEN NOT MATCHED BY SOURCE THEN UPDATE SET ...
WHEN NOT MATCHED THEN INSERT ...
```

Which means:
 * Filtering condition not only is applicable to "matched" but also can be applied to the other two cases ("not matched by source/target").
 * All three merge situation clauses, no matter having additional filtering condition or not, can be repeated as long as having valid syntax (parser doesn't care about unreachable case)

**Comment by pykenny** [comment]
`NOT MATCHED BY SOURCE` and `MATCHED` cases allows delete and update (but not insert) operations, while `NOT MATCHED (BY TARGET)` only allows insert operation. I think that constraint is absent in current implementation as well.

File: sqlalchemy_bigquery/merge.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
+import textwrap
```
[the remark this answers is not in the document]
Comment (munro) [reply sqlalchemy_bigquery/merge.py]: Added!

File: sqlalchemy_bigquery/merge.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+import sqlalchemy as sa
+from sqlalchemy.ext.compiler import compiles
+from sqlalchemy.sql.compiler import SQLCompiler
+from sqlalchemy.sql.expression import ClauseElement, ColumnElement
+from sqlalchemy.sql.selectable import Subquery  # noqa
+from sqlalchemy.sql.selectable import Alias, Select, TableClause
+
+
+class MergeThen:
+    ...
```
[the remark this answers is not in the document]
Comment (munro) [reply sqlalchemy_bigquery/merge.py]: Fixed!

File: sqlalchemy_bigquery/merge.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
+import textwrap
+import textwrap
+from dataclasses import dataclass
```
[the remark this answers is not in the document]
Comment (munro) [reply sqlalchemy_bigquery/merge.py:8]: https://github.com/googleapis/python-bigquery-sqlalchemy/blob/f5663718c4727a55157e8f0cdd95c371311498aa/setup.py#L98

Ah went to go add it, but I see it now requires `python >= 3.7` 😄

File: tests/unit/test_merge.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
+import sqlalchemy as sa
```
[the remark this answers is not in the document]
Comment (munro) [reply tests/unit/test_merge.py]: Fixed!

**Comment by munro** [comment]
@pykenny Oh wow, thank you, I didn't even know you could have multiple of the same when conditions.  I read the documentation [1] and made it match.

I don't like the syntax (currently).

I really wanted to avoid 'piping', but IDK, it may be nicer...?

```python
merge = Merge(
    target=target,
    source=source,
    on=into.c.b == using.c.b
)

merge = merge.when_matched(target.foo > 10).then_insert({"foo": source.foo})
merge = merge.when_matched_not_matched_by_source().then_delete()
```

... it does kinda look nice. 😄 

Also contemplating a 'partial' piping syntax:


```python
Merge(
    target=target,
    source=source,
    on=into.c.b == using.c.b,
    when=[
        Merge.WhenMatched(target.foo > 10).then_insert({"foo": source.foo})
    ]
)
```




[1] https://cloud.google.com/bigquery/docs/reference/standard-sql/dml-syntax#merge_statement

File: sqlalchemy_bigquery/merge.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+        self.when: Tuple[WhenBase] = tuple()
+
+    def when_matched(self, condition: MergeConditionType = None):
+        return MergeWhenMatched(self, condition)
+
+    def when_matched_not_matched_by_target(self, condition: MergeConditionType = None):
+        return MergeWhenNotMatchedByTarget(self, condition)
+
+    def when_matched_not_matched_by_source(self, condition: MergeConditionType = None):
+        return MergeWhenNotMatchedBySource(self, condition)
```
Comment (pykenny) [inline sqlalchemy_bigquery/merge.py:61]: Shouldn't they be `when_not_matched_by_target` and `when_not_matched_by_source`? 🤔

File: sqlalchemy_bigquery/merge.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    self: WhenNotMatchedByTarget, compiler: SQLCompiler, **kwargs
+):
+    return "WHEN NOT MATCHED BY TARGET" + _compile_when(self, compiler, **kwargs)
+
+
+@compiles(WhenNotMatchedBySource)
+def _compile_when_not_matched_by_source(
+    self: WhenNotMatchedBySource, compiler: SQLCompiler, **kwargs
+):
+    return "WHEN MATCHED NOT MATCHED BY SOURCE" + _compile_when(
```
Comment (pykenny) [inline sqlalchemy_bigquery/merge.py:187]: `WHEN MATCHED NOT MATCHED BY SOURCE` -> `WHEN NOT MATCHED`

File: sqlalchemy_bigquery/merge.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+from sqlalchemy.ext.compiler import compiles
+from sqlalchemy.sql.compiler import SQLCompiler
+from sqlalchemy.sql.expression import ClauseElement, ColumnElement
+from sqlalchemy.sql.selectable import Subquery  # noqa
+from sqlalchemy.sql.selectable import Alias, Select, TableClause
+
+MergeConditionType = Optional[ColumnElement[sa.Boolean]]
+
+
+class Merge(ClauseElement):
```
Comment (pykenny) [inline sqlalchemy_bigquery/merge.py:22]: `ClauseElement` instance is not executable, so passing a `Merge` instance to `Connection.execute()` doesn't work.

```python
updates = {"id": source.id, "name": source.name}
stmt = (
  Merge(target, source, on=target.id == source.id)
  .when_matched()
  .then_update(updates)
  .when_not_matched_by_target()
  .then_insert(updates)
)
connection.execute(stmt)
```

How about inheriting [`UpdateBase`](https://docs.sqlalchemy.org/en/14/core/dml.html#sqlalchemy.sql.expression.UpdateBase)?

**Comment by pykenny** [comment]
@munro Would it be possible to make it support ORM models as well? Users may want to write this:

```python
from sqlalchemy.orm import aliased
...

alias_dest = aliased(MyModelDest)
alias_source = aliased(MyModelSource)

Update(
    alias_dest,
    alias_source,
    on=...,
).when_matched().then_update(...)
```

**Comment by meredithslota** [comment]
Unfortunately, we can't accept contributions without a signed CLA. This has been open for a while with a couple comments to this effect, and while I can tell it would be helpful, we can't accept it as-is. See https://github.com/googleapis/python-bigquery-sqlalchemy/pull/429#issuecomment-1075198520 and our [CONTRIBUTING guide](https://github.com/googleapis/python-bigquery-sqlalchemy/blob/main/CONTRIBUTING.rst) for more info.

**Comment by munro** [comment]
> @munro Thanks so much for the contribution. I hope to review soon. Unfortunately, I cannot merge this without the account associated with your commit (I see a kw.com email) signing the Google CLA at https://cla.developers.google.com/clas

@meredithslota I just signed the CLA! [1] Signed it because I see activity in the PR #428 

[1] https://cla.developers.google.com/clas/new?domain=DOMAIN_GOOGLE&kind=KIND_INDIVIDUAL

**Comment by JacobHayes** [comment]
Can this be reopened or resubmitted now that the CLA is signed?

**Pull request closed without merging**